### PR TITLE
feat: ✨ Implement `CommitmentVerifier` for `FileKeyVerifier` used to verify key proofs in Proofs Dealer pallet

### PR DESCRIPTION
**Pull request by ffarall**
In this PR:

- [x] Create new struct `FileKeyVerifier` that implements `CommitmentVerifier`, to provide a way of verifying file key proofs, where hash challenges are interpreted as numbers to which we take the module of the number of chunks, and the result is the challenged chunk.
- [x] Abstract `CommitmentVerifier` trait over the `Commitment` and `Challenge` types.
- [x] Initiate refactor of `support/primitives` and `support/traits` folders. ***ATTENTION REVIEWER!!!*** The code in the aforementioned two folders, should be _equal_ to the code in `primitives/forest-verifier` and `primitives/traits`, respectively (except maybe in names).

### CODE REVIEW

File: primitives/file-key-verifier/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        challenges: &[Self::Challenge],
+        proof: &Self::Proof,
+    ) -> Result<Vec<Self::Challenge>, DispatchError> {
+        // Check that `challenges` is not empty.
+        if challenges.is_empty() {
+            return Err("No challenges provided.".into());
+        }
+
+        // Construct file key from the fields in the proof.
+        let file_key = T::Hash::hash(
```
Comment (links234) [inline primitives/file-key-verifier/src/lib.rs:70]: This code will be reused. A good reason for `FileMetadata::file_key(&self)`.

File: primitives/file-key-verifier/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                &proof.fingerprint.encode(),
+            ]
+            .into_iter()
+            .flatten()
+            .cloned()
+            .collect::<Vec<u8>>(),
+        );
+
+        // Check that the number of challenges is proportional to the size of the file.
+        let mut chunks_to_check = proof.size / SIZE_TO_CHALLENGES;
```
Comment (links234) [inline primitives/file-key-verifier/src/lib.rs:84]: This looks like it's going to be reused too, i.e. `FileMetadata::chunks_to_challenge(&self);`.

File: primitives/file-key-verifier/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        // Initialise vector of proven challenges. We use a `BTreeSet` to ensure that the items are unique.
+        let mut proven_challenges = BTreeSet::new();
+        let mut challenges_iter = challenges.iter();
+
+        // Iterate over the challenges, compute the modulo of the challenged hashes with the number of chunks in the file,
+        // and check if the resulting leaf is in the proof.
+        while let Some(challenge) = challenges_iter.next() {
+            // Calculate the chunks of the file based on its size.
+            let mut chunks = proof.size / CHUNK_SIZE;
```
Comment (links234) [inline primitives/file-key-verifier/src/lib.rs:130]: `FileMetadata::chunks(&self);`

File: primitives/file-key-verifier/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            if proof.size % CHUNK_SIZE != 0 {
+                chunks += 1;
+            }
+
+            // Calculate the modulo of the challenge with the number of chunks in the file.
+            // The challenge is a big endian 32 byte array.
+            let challenged_chunk = BigUint::from_bytes_be(challenge.as_ref()) % chunks;
```
Comment (links234) [inline primitives/file-key-verifier/src/lib.rs:137]: `FileMetadata::challenged_chunk(&self, challenge: [32; u8]);`

File: primitives/forest-verifier/src/tests.rs [hunk 5]
```diff
@@ -0,0 +1,1375 @@
...
+    (memdb, root, file_keys)
+}
+
+/// Chunk a file into [`CHUNK_SIZE`] byte chunks and store them in a Merkle Patricia Trie.
+///
+/// The trie is stored in a [`MemoryDB`] and the [`Root`] is returned.
+///
+/// TODO: make this function fetch data from storage using Storage trait.
+/// _It is assumed that the file is located in the same directory as the executable._
+pub fn merklise_file<T: TrieLayout>(file_path: &str) -> (MemoryDB<T::Hash>, HashT<T>) {
```
Comment (links234) [inline primitives/forest-verifier/src/tests.rs:196]: This should use `InMemoryFileDataTrie` and maybe moved inside `FileDataTrie` trait.

File: primitives/forest-verifier/src/tests.rs [hunk 6]
```diff
@@ -0,0 +1,1375 @@
...
+    let file_path = FILES_BASE_PATH.to_owned() + file_path;
+    let mut file = std::fs::File::open(file_path).unwrap();
+    let mut buf = [0; CHUNK_SIZE];
+    let mut chunks = Vec::new();
+
+    // create list of key-value pairs consisting of chunk metadata and chunk data
+    loop {
```
Comment (links234) [inline primitives/forest-verifier/src/tests.rs:203]: This is used for testing but if you decide to move it to `FileDataTrie`,  it should not read everything in memory but construct it as you go. See [this](https://github.com/Moonsong-Labs/storage-hub/pull/72/files#diff-f5b58703c6605ce51568c53912a42386c530956f2f22e15c0a4bcaa4bdfbdf65R89).

**Review by links234 — APPROVED** [review]
Great work! Mainly I left some refactor suggestions which I will have to do in my PR anyway.

File: primitives/forest-verifier/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+mod tests;
+
+/// A struct that implements the `CommitmentVerifier` trait, where the commitment
+/// is a Merkle Patricia Trie root hash and the response to a challenge is given
+/// by either the exact key or the next and previous keys in the trie.
+pub struct ForestVerifier<T: TrieLayout, const H_LENGTH: usize>
+where
+    <T::Hash as sp_core::Hasher>::Out: for<'a> TryFrom<&'a [u8; H_LENGTH]>,
+{
+    pub _phantom: core::marker::PhantomData<T>,
```
Comment (snowmead) [inline primitives/forest-verifier/src/lib.rs:19]: nipick: normally when there are no fields in a struct, you can have a tuple instead - `ForestVerifier(core::marker::PhantomData<T>)`.

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -103,6 +103,15 @@ pub type Executive = frame_executive::Executive<
...
 >;
 
+/// The file chunk size in bytes. This is the size of the leaf nodes in the Merkle
+/// Patricia Trie that is constructed for each file.
+/// Each chunk is 1 kB.
+pub const FILE_CHUNK_SIZE: u64 = 2u64.pow(10);
+
+/// The number of challenges for a file, depending on the size of the file.
+/// For every 512 kB, there is a challenge.
+pub const FILE_SIZE_TO_CHALLENGES: u64 = 2u64.pow(25);
```
[the remark this answers is not in the document]
Comment (snowmead) [reply runtime/src/lib.rs:113]: I don't really understand this constant. Is it supposed to be a ratio? And why don't we just specify a static number (same for `FILE_CHUNK_SIZE`)?

File: pallets/file-system/src/utils.rs [hunk 9]
```diff
@@ -340,6 +340,16 @@ where
             &key_proof,
         )?;
 
+        // TODO: Generate challenges for the key proof properly.
```
[the remark this answers is not in the document]
Comment (ffarall) [reply pallets/file-system/src/utils.rs:343]: Thanks for catching this, it must have duplicated in the merge

**Pull request merged**